main/src/ui/manage_accounts/certificate_pinning_dialog.vala

```diff
@@ -0,0 +1,70 @@
+using Gdk;
+using Gee;
+using Gtk;
+using Markup;
+using Gcr;
+
+using Dino.Entities;
+
+namespace Dino.Ui.ManageAccounts {
+
+[GtkTemplate (ui = "/im/dino/Dino/manage_accounts/certificate_pinning_dialog.ui")]
+public class CertificatePinningDialog : Gtk.Dialog {
+
+    [GtkChild] public CertificateWidget certificate_widget;
+    [GtkChild] public Label info_label;
+    [GtkChild] public Button pin_certificate_button;
+
+    private Database db;
+    private Account account;
+    private TlsCertificate? server_certificate;
+    private bool do_trust;
+
+    construct {
+        pin_certificate_button.clicked.connect(on_pin_certificate);
+    }
+
+    public CertificatePinningDialog(Database db, Account account, TlsCertificate? server_certificate, bool used_pinned_certificate) {
+        Object(use_header_bar : 1);
+        this.db = db;
+        this.account = account;
+        this.server_certificate = server_certificate;
+        if (server_certificate != null) {
+            Certificate gcr_certificate = new SimpleCertificate(server_certificate.certificate.data);
+            certificate_widget.set_certificate(gcr_certificate);
+        } else {
+           info_label.label = "";
+           pin_certificate_button.sensitive = false;
+           return;
+        }
+        if (!used_pinned_certificate) {
+           // only show certificate, no pinning needed
```

Comment:
It is currently ignoring errors for a given certificate rather than pinning it, but if we move to actually pinning a key/certificate, we would probably want to be able to pin a valid one.